rfcs/20181217-tf2-random-numbers.md

```diff
@@ -0,0 +1,206 @@
+*   If you specify just the op seed, we use <code>[DEFAULT_GLOBAL_SEED](https://github.com/tensorflow/tensorflow/blob/3eb7616b5459aec3dabaa4152a00de14a1fa0914/tensorflow/python/framework/random_seed.py#L29)</code> for the global seed so you get deterministic behavior.
+
+## Design Proposal
+
+The following represents the desired end-state, and doesn't go into detail about transitioning from our current stateful ops:
+
+```python
+# random.py
+
+# A seed for random ops (stateful and stateless) will always be 256 
+# bits, all of which will be sent to the C++ code. The actual C++ 
+# implementation of some algorithms may only use a lower part of the bits.
+# *QUESTION*: Is 256 a good number? 
+  
+@tf_export("random.non_deterministic_seed")
+def non_deterministic_seed():  # returns an integer
+  # *QUESTION*: Is this pure Python or an op?
+  # PROPOSAL: Op; need only be implemented on CPU
+
+# *QUESTION*: Should this be public?
+def create_rng_state(seed, algorithm=None):
+  # seed must be an integer or stateless seed, never None
+  # algorithm=None -> auto-select
+  # Returns a 1-D tensor "rng_state" with:
+  # * rng_state[0] is a value that identifies the RNG algorithm;
+  # * rng_state[1:] holds the RNG state itself (size dependent on the 
+  #                 algorithm).
+
+@tf_export("random.Generator")
+class Generator(Checkpointable):
+  def __init__(self, copy_from=None, seed=None, algorithm=None):
+    if copy_from is None:
+      if seed is None:
+        seed = non_deterministic_seed()
+      self._state_var = tf.Variable(create_rng_state(seed, algorithm))
+    else:
+      assert seed is None
+      self._state_var = tf.Variable(copy_from.state)
+
+  @property
+  def state(self):
+    return self._state_var
+
+  @property
+  def algorithm(self):
+    return self._state_var[0]
+
+  # The following functions return a tensor and as a side effect update 
+  # self._state_var.
+  def uniform(self, shape, minval=0, maxval=None, dtype=tf.float32, name=None):
+  def normal(self, shape, mean=0.0, stddev=1.0, dtype=tf.float32, name=None):
+  def make_seeds(self, shape=()):  # generates seeds for stateless random ops
+  def make_generators(self, count=1, name=None):
+    # Returns a list of `count` independent `Generator` objects
+  # ...
+
+op_generator = Generator()
+
+@tf_export("random.reset_global_generator")
+def reset_global_generator(seed, algorithm=None):
+  global op_generator
+  if algorithm is None:
+    algorithm = op_generator.algorithm # preserve the old algorithm
+  op_generator = Generator(seed=seed, algorithm=algorithm)
+
+@tf_export("random.set_global_generator")
+def set_global_generator(generator):
```

Comment:
It also seems a bit inconsistent to have the world `global` in the method name while the internal variable is named `op_generator`. Do we need the word `global` here if there is no local generator?

Reply:
The internal name `op_generator` is not part of the API and is actually called `global_generator` in the implementation (I should change this in the RFC). Whether the API name should contain the word "global" is up for debate.

Reply:
I'll leave "global" in the name since "the global generator" (a singleton) is a phrase we use often in our conversations.